res/controllers/Numark-Mixtrack-Platinum-scripts.js

```diff
@@ -0,0 +1,1467 @@
+    bpmArray = MixtrackPlatinum.encodeNumToArray(bpm);
+    bpmArray.shift();
+    bpmArray.shift();
+
+    var bytePrefix = [0xF0, 0x00, 0x20, 0x7F, deck, 0x01];
+    var bytePostfix = [0xF7];
+    var byteArray = bytePrefix.concat(bpmArray, bytePostfix);
+    midi.sendSysexMsg(byteArray, byteArray.length);
+};
+
+MixtrackPlatinum.bpmCallback = function(value, group, control) {
+    var midi_chan = MixtrackPlatinum.channelMap[group];
+    MixtrackPlatinum.screenBpm(midi_chan + 1, Math.round(value * 100));
+};
+
+MixtrackPlatinum.channelMap = {
+    "[Channel1]": 0x00,
+    "[Channel2]": 0x01,
+    "[Channel3]": 0x02,
+    "[Channel4]": 0x03,
+};
+
+MixtrackPlatinum.elapsedToggle = function (channel, control, value, status, group) {
+    if (value != 0x7F) return;
+
+    var current_setting = engine.getValue('[Controls]', 'ShowDurationRemaining');
+    if (current_setting === 0) {
+        // currently showing elapsed, set to remaining
+        engine.setValue('[Controls]', 'ShowDurationRemaining', 1);
+    } else if (current_setting === 1) {
+        // currently showing remaining, set to elapsed
+        engine.setValue('[Controls]', 'ShowDurationRemaining', 0);
+    } else {
+        // currently showing both (that means we are showing remaining, set to elapsed
+        engine.setValue('[Controls]', 'ShowDurationRemaining', 0);
+    }
+};
+
+MixtrackPlatinum.timeElapsedCallback = function(value, group, control) {
+    // 0 = elapsed
+    // 1 = remaining
+    // 2 = both (we ignore this as the controller can't show both)
+    var on_off;
+    if (value === 0) {
+        // show elapsed
+        on_off = 0x00;
+    } else if (value === 1) {
+        // show remaining
+        on_off = 0x7F;
+    } else {
+        // both, ignore the event
+        return;
+    }
+
+    // update all 4 decks on the controller
+    midi.sendShortMsg(0x90, 0x46, on_off);
+    midi.sendShortMsg(0x91, 0x46, on_off);
+    midi.sendShortMsg(0x92, 0x46, on_off);
+    midi.sendShortMsg(0x93, 0x46, on_off);
+};
+
+MixtrackPlatinum.timeMs = function(deck, position, duration) {
+    return Math.round(duration * position * 1000);
+};
+
+MixtrackPlatinum.positionCallback = function(playposition, group, control) {
+    var midi_chan = MixtrackPlatinum.channelMap[group];
+    // the controller appears to expect a value in the range of 0-52
+    // representing the position of the track. Here we send a message to the
+    // controller to update the position display with our current position.
+    var pos = Math.round(playposition * 52);
+    if (pos < 0) pos = 0;
+    midi.sendShortMsg(0xB0 | midi_chan, 0x3F, pos);
+
+    // update duration if necessary
+    var duration = engine.getValue(group, 'duration');
+    MixtrackPlatinum.screenDuration(midi_chan + 1, duration * 1000);
+
+    // update the time display
+    var time = MixtrackPlatinum.timeMs(midi_chan + 1, playposition, duration);
+    MixtrackPlatinum.screenTime(midi_chan + 1, time);
```

Comment:
How does this handle showing time remaining? Or do you just need to send the controller firmware the time elapsed and duration, then it calculates remaining if you set it to show remaining?

Reply:
Right. We tell the controller the duration and we send the current time elapsed. The controller shows remaining/elapsed depending on which one it has been told to do.

Reply:
Okay, please add a comment explaining that.